### PR TITLE
Add simple starting point to write Property based test

### PR DESCRIPTION
**Pull request by joemphilips**
1. Currently, it is only for very basic stuff (Address and TX serialization)
2. It is quite slow (I'm not sure how to make it fast, using different seed generation algo for KeyGenerator might work.)

Any feedback appreciated.
As a future work, I am planning to implement something like `ConsensusAcceptableTransactionGenerator` and `MempoolAcceptableTransactionGenerator` .

### CODE REVIEW

**Comment by NicolasDorier** [comment]
Cool, can you rebase? you are adding a file that I deleted (Readme.txt)

File: NBitcoin.Tests/Generators/ScriptGenerator.cs [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        return multiSignatureScriptSig();
+      }
+
+      throw new Exception("Unknown Script PubKey");
+    }
+
+    #endregion
+
+    #region scriptPubKey
+    public static Gen<Script> p2pkhScriptPubKey() =>
```
Comment (NicolasDorier) [inline NBitcoin.Tests/Generators/ScriptGenerator.cs]: `p2pkhScriptPubKey` => `P2PKHScriptPubKey`

**Comment by NicolasDorier** [comment]
Can you fix the cases ? In C#, we use PascalCase.

**Comment by joemphilips** [comment]
Looks good now?

Where should I put explanation about these tests? The explanation is like following (BTY, I'd like to hear your thoughts about these categories if you have any opinion.)

```
"PropertyTest" are Property based test. i.e. It generates some test data automatically and test against all those data.
For PropertyTest, we split tests into following categories.
(These are an exactly same with the one described in ref: https://fsharpforfunandprofit.com/posts/property-based-testing-2/)

1. Commutativity
2. BidirectionalConversion
3. Immutability
4. Idempotence
5. Induction
6. Verification
7. Oracle
```

**Comment by joemphilips** [comment]
done.

**Comment by NicolasDorier** [comment]
It makes me think that I probably should add a BenchmarkDotnet project to identify where NBitcoin is slow.

**Comment by NicolasDorier** [comment]
I fixed the perf issue on c8dd9ae58624c7fabeb3a257560bb84dd3b34a1b there was too much amount above 21M.

**Comment by NicolasDorier** [comment]
Can you take a look at `Assert.True(tx.HasWitness);` I tried removing the `MaxTest = 10`, then things start breaking as it seems your generator is generating non segwit tx

**Comment by NicolasDorier** [comment]
Found the issue

**Comment by NicolasDorier** [comment]
2fb4f3416ed77fd83681fc54920cc22cf3d002b3

**Pull request merged**